<html><table>
<tr><td>term:<td> alanine scanning
<tr><td>id:<td> <a name="MI:0005">MI:0005
<tr><td>definition:<td> This approach is used to identify the residues that are involved in an interaction. Several variants of the native protein are prepared by mutating each residue of interest to an alanine. The mutated proteins are expressed, tested for proper folding and probed in the binding assay.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> complete sequence: full identification by sequencing
<tr><td>id:<td> <a name="MI:0056">MI:0056
<tr><td>definition:<td> The DNA to be sequenced is used as template for the in vitro synthesis, by DNA polymerase, of a set of partial replicas, all beginning at the same place, but terminating at different points along the DNA chain. The key to this method is the use of dideoxyribonucleoside triphosphates in which the deoxyribose 3'-OH group present in normal nucleotides is missing; when such a modified nucleotide is incorporated into a DNA chain, it blocks the elongation of the chain. To determine the full sequence, the four different chain-terminating dideoxyribonucleosides are used in competition with an excess of deoxyribonucleosides in separate DNA synthesis reactions on the same DNA template. When the products of these four reactions are analysed by electrophoresis in four parallel lanes of a denaturing polyacrylamide gel, the DNA sequence can be derived. Every lane displays a family of DNA fragments of different lengths, reflecting the different sites at which a specific residue occurs in the original DNA.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> correlated mutations
<tr><td>id:<td> <a name="MI:0026">MI:0026
<tr><td>definition:<td> Pairs of multiple alignments of orthologous sequences are used to identify potential interacting partners as proteins that show covariation of their residue identities between different species. Proteins displaying inter-protein correlated mutations during evolution are likely to be interacting proteins due to co-adapted evolution of their protein interacting interfaces.
<tr><td>definition_reference:<td> PMID:11933068
<tr><td colspan=2><hr><tr><tr><td>term:<td> deletion analysis
<tr><td>id:<td> <a name="MI:0033">MI:0033
<tr><td>definition:<td> In this approach, once a protein is demonstrated to participate in an interaction, several deletion derivatives are produced and tested in the binding assay to identify the minimal fragment (domain) that can still support the interaction.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> docking
<tr><td>id:<td> <a name="MI:0035">MI:0035
<tr><td>definition:<td> Predicts the structure of a protein-protein complex from the unbound structures of its components.  The initial approach in the majority of docking procedures is based largely on the 'rigid-body' assumption, whereby the proteins are treated as solid objects. Initial scoring of a complex is based on geometric fit or surface complementarity. This generally requires some knowledge of the binding site to limit the number of solutions.
<tr><td>definition_reference:<td> PMID:9631301
<tr><td colspan=2><hr><tr><tr><td>term:<td> domain profile pairs
<tr><td>id:<td> <a name="MI:0037">MI:0037
<tr><td>definition:<td> This approach use a protein interaction network of a given organism to infer interaction in another organism using information about the interacting region. The regions or domains involved in interactions are clustered if they share sequence similarity and have common interacting partners. The resulting domain profiles are then used to screen the proteome of another organism and domain-domain interactions are inferred. Ultimately, an inferred protein interaction map is built in this second organism.
<tr><td>definition_reference:<td> PMID:11473021
<tr><td colspan=2><hr><tr><tr><td>term:<td> electron resonance
<tr><td>id:<td> <a name="MI:0043">MI:0043
<tr><td>definition:<td> A form of spectroscopy in which the absorption of microwave by a sample in a strong magnetic field is used to study atoms or molecules with unpaired electrons.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> endor: electron nuclear double resonance
<tr><td>id:<td> <a name="MI:0041">MI:0041
<tr><td>definition:<td> A combination of NMR and EPR. The lines in the EPR spectrum that are caused by coupling of an unpaired electron nearby nuclei change in intensity when these nuclei are excited at their NMR frequency.
<tr><td>definition_reference:<td> PMID:11817959
<tr><td>definition_reference:<td> PMID:11988476
<tr><td colspan=2><hr><tr><tr><td>term:<td> epr: electron paramagnetic resonance
<tr><td>id:<td> <a name="MI:0042">MI:0042
<tr><td>definition:<td> EPR (also called ESR, Electron Spin Resonance) spectroscopy is analogous to NMR, but is based on the excitation of unpaired electrons instead of nuclei. Unpaired (single) electron are only found in radicals and some metal ions (paramagnetic species); the EPR spectrum provides information about the environment and mobility of the paramagnetic species. The magnetic interaction of two paramagnetic centers in a protein can be used to calculate the distance between them; this allows studies of the movements and interactions of protein segments. In proteins without any intrinsic unpaired electrons it is possible to attach a radical probe (spin label). Stable nitroxide radicals can be bound to amino acid residues, in analogy with fluorescent probes. In combination with site directed mutagenesis this method is used in particular to study structure and assembly of membrane proteins, by measuring with EPR whether an amino acid is in a polar or non polar environment.
<tr><td>definition_reference:<td> PMID:11817959
<tr><td colspan=2><hr><tr><tr><td>term:<td> feature detection
<tr><td>id:<td> <a name="MI:0003">MI:0003
<tr><td>definition:<td> Method to determine the features of the proteins involved in the interaction.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> interface predictor: neural network on interface properties
<tr><td>id:<td> <a name="MI:0076">MI:0076
<tr><td>definition:<td> Neural networks are trained on the properties of residues belonging to a cluster of residues that are neighbors in space on protein surface. The predictor permits the inference of the residues that are likely to be on an interaction interface.
<tr><td>definition_reference:<td> PMID:11455607
<tr><td>definition_reference:<td> PMID:11874449
<tr><td colspan=2><hr><tr><tr><td>term:<td> mobility shift
<tr><td>id:<td> <a name="MI:0070">MI:0070
<tr><td>definition:<td> Protein modifications can be identified by gel electrophoresis since any change in the mass and/or the charge of the protein can alter its mobility in PAGE. Although this method does not allow the unequivocally identification of the type of modification that has caused the shift, it is possible, by combining this approach with more direct methods, to correlate the extent of the shift to a specific modification.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> modified residue ms: mass detection of residue modification
<tr><td>id:<td> <a name="MI:0068">MI:0068
<tr><td>definition:<td> Mass spectrometry can be used to characterise chemical modifications within peptides. One approach consists in the observation of a mass difference when a sample is treated with an enzyme that can specifically remove a peptide modification, for instance a phosphatase. The mass difference corresponds to the mass of the chemical group covalently linked to a residue. Such experiments carried out with a MALDI-TOF (Matrix-assisted laser desorption ionization time-of-flight ) do not allow the mapping of the modification site within the sequence, whereas any tandem mass spectrometer (LC MS/MS Liquid Chromatography Tandem Mass Spectrometry, nanoESI MS/MS nanoElectrospray Ionisation tandem mass spectrometry, FTMS Fourier Transform mass spectrometry) provide such information. A second approach consists of the direct mass measurement of the ionized chemical group dissociated from the residue within a tandem mass spectrometer. Both approaches need a prior enrichment of the modified peptide population in the samples with IMAC (Immobilized Metal Affinity Chromatography)or specific anti-modification antibodies.
<tr><td>definition_reference:<td> PMID:11395414
<tr><td colspan=2><hr><tr><tr><td>term:<td> monoclonal antibody
<tr><td>id:<td> <a name="MI:0072">MI:0072
<tr><td>definition:<td> monoclonal antibodies are monospecific antibodies produced in the supernatant of a cell line obtained by fusing a lymphocyte B to a myeloma cell line or selected by phage display technology.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> mutation analysis
<tr><td>id:<td> <a name="MI:0074">MI:0074
<tr><td>definition:<td> Several mutant proteins are produced by random or directed techniques and assayed for their ability to support binding. Mutants defective in binding are tested for correct folding to pinpoint the residues that are directly involved in binding.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> nmr: nuclear magnetic resonance
<tr><td>id:<td> <a name="MI:0077">MI:0077
<tr><td>definition:<td> NMR requires a small volume of concentrated protein solution that is placed in a strong magnetic field. Certain atomic nuclei, and in particular hydrogen, have a magnetic moment or spin; that is, they have an intrinsic magnetisation, like a bar magnet. The spin aligns along the strong magnetic field, but can be changed to a misaligned excited state in response to applied radio frequency (RF) pulses of electromagnetic radiation. When the excited hydrogen nuclei relax to their aligned state, they emit RF radiation, which can be measured and displayed as a spectrum. The nature of the emitted radiation depends on the environment of each hydrogen nucleus, and if one nucleus is excited, it will influence the absorption and emission of radiation by other nuclei that lie close to it. It is consequently possible, by an ingenious elaboration of the basic NMR technique known as two-dimensional NMR, to distinguish the signals from hydrogen nuclei in different amino acid residues and to identify and measure the small shifts in these signals that occur when these hydrogen nuclei lie close enough to interact: the size of such a shift reveals the distance between the interacting pair of hydrogen atoms. In this way NMR can give information about the distances between the parts of the protein molecule. NMR provides information about interacting atoms thereby permitting to obtain information about protein structure and protein-protein interaction.
<tr><td>definition_reference:<td> PMID:12120505
<tr><td colspan=2><hr><tr><tr><td>term:<td> polyclonal antibody
<tr><td>id:<td> <a name="MI:0086">MI:0086
<tr><td>definition:<td> Polyclonal antibodies are a mixture of different antibodies that represent the immune response, normally in an experimental animal, to any given antigen.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> protein sequence_ms: de novo protein sequencing by mass spectrometry
<tr><td>id:<td> <a name="MI:0032">MI:0032
<tr><td>definition:<td> The strategy to determine the complete amino acid sequence of a protein by mass spectrometry relies on the generation of a nested set of fragments differing by one amino acid. This permits to reveal the identity of the residue that has been removed at each degradation step by measuring the mass difference of fragments differing of one residue. Peptide fragments can be obtained by protease treatment combined with the fragmentation promoted by collision (or other methods) within a tandem mass spectrometer. This approach can be carried out with LC MS/MS (Liquid Chromatography Tandem Mass Spectrometry), nanoESI MS/MS (nanoElectrospray Ionisation tandem mass spectrometry), or FTMS (Fourier Transform mass spectrometry) instruments.
<tr><td>definition_reference:<td> PMID:10984529
<tr><td colspan=2><hr><tr><tr><td>term:<td> protein staining
<tr><td>id:<td> <a name="MI:0094">MI:0094
<tr><td>definition:<td> A wide range of dyes have been used over the years to visualise proteins in polyacrylamide gels - Coomasie Blue and silver-staining being two classical methods. Fluorescent dyes such as Nile Red and SYPRO Orange are now increasingly used due to their superior dynamic range. Use of non-denaturing gels can allow visualisation of protein protein interactions. Several dyes can be used to specifically indicate residue modification, however this methodology will give no information as the number of residues modified or their position within the protein sequence. Examples include the use of acid fuscian or the fluorescent dansyl hydrazine to show protein glycosylation.
<tr><td>definition_reference:<td> PMID:12015990
<tr><td colspan=2><hr><tr><tr><td>term:<td> surface patches
<tr><td>id:<td> <a name="MI:0106">MI:0106
<tr><td>definition:<td> Surface patches are built using 6 criteria: solvation potential, residue interface propensity, hydrophobicity, planarity, protrusion and accessible surface area. Protein structures having similar patches are likely to have the same interactions.
<tr><td>definition_reference:<td> PMID:9299343
<tr><td colspan=2><hr><tr><tr><td>term:<td> western blot
<tr><td>id:<td> <a name="MI:0113">MI:0113
<tr><td>definition:<td> Western blot is a procedure to identify and quantify proteins. A mixture of protein is first submitted to an electrophoresis in denaturant condition and then electro-transferred from the gel to a membrane. The membrane is then incubated with a primary antibody specific for a given protein or a specific residue modification in the sample under analysis. A secondary antibody, radiolabelled or fused to a chromogenic enzyme, targets the first antibody and allows the visualisation of the protein band on the membrane.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> x-ray: x-ray crystallography
<tr><td>id:<td> <a name="MI:0114">MI:0114
<tr><td>definition:<td> X-rays have a wavelength, typically around 0.1 nm (the diameter of a hydrogen atom). If a narrow parallel beam of X-rays is directed at a sample of a pure protein, most of the X-rays will pass straight through it. A small fraction, however, will be scattered by the atoms in the sample. If the sample is a well-ordered crystal, the scattered waves will reinforce one another at certain points and will appear as diffraction spots when the X-rays are recorded by a suitable detector. The position and intensity of each spot in the X-ray diffraction pattern contain information about the position and nature of the atoms in the crystal. The three-dimensional structure of a large molecule can be deduced from the electron-density map of its crystal. In recent years X-ray diffraction analysis has become increasingly automated, and now the slowest step is likely to be the production of suitable protein crystals. This requires high concentration of very pure protein and empirical searching for the proper crystallisation conditions.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr></table></html>