<html><table>
<tr><td>term:<td> antibody detection: identification by antibody
<tr><td>id:<td> <a name="MI:0421">MI:0421
<tr><td>definition:<td> antibody mediated identification
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> edman degradation
<tr><td>id:<td> <a name="MI:0039">MI:0039
<tr><td>definition:<td> In this procedure the N-terminus amino acid is cleaved from a polypeptide and identified by high-pressure liquid chromatography. The cycle is repeated on the evershortening polypeptide until all the residues are identified. On average only 20-30 consecutive cycles can be performed and lead to amino acid identification. Longer polypeptides or full length proteins must be cleaved by specific protease before Edman degradation and their sequences built by fragment overlapping.
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> elisa: enzyme linked immunosorbent assay
<tr><td>id:<td> <a name="MI:0411">MI:0411
<tr><td>definition:<td> Following non-covalent binding of a purified primary ligand to a solid phase, a blocking reagent is added to prevent any non-specific binding. A specific antigen is then allowed to bind to the primary ligand. Unbound antigen is removed by washing and a secondary antibody conjugated to an enzyme (e.g. horseradish peroxidase) is added. Following a washing step to remove unbound secondary ligand, the extent to which a chromogenic substrate (e.g. 3,3', 5,5' tetramethyl benzidine chromogen [TMB]) is converted to a soluble coloured product by the conjugated enzyme in a given time is determined by spectrophotometry using a standard microplate absorbance reader. A similar type of approach can be utilized to detect enzymatic activities. The substrate, attached to a solid phase is incubated in the presence of the enzyme and the enzymatic modification is monitored by an antibody that is specific for the modified substrate (for instance a phosphorylated protein).
<tr><td>definition_reference:<td> PMID:11906746
<tr><td colspan=2><hr><tr><tr><td>term:<td> fingerprinting: peptide massfingerprinting
<tr><td>id:<td> <a name="MI:0082">MI:0082
<tr><td>definition:<td> This approach leads to protein identification by matching peptide masses, as measured by mass spectrometry, to the ones calculated from in silico fragmentation of a protein sequence database. A peptide mixture from a tryptic digest is analysed by MALDI-MS (Matrix-assisted laser desorption ionization mass spectrometry). The list of peptide masses obtained by MALDI MS is automatically compared to the calculated masses of the predicted peptide fragments for each entry in the database. High mass accuracy is very important in order to obtain a statistically significant and unambiguous match This method is best applied to completely sequenced genomes and well characterised proteomes. However, depending on the data quality, proteins that are highly homologous to already characterised proteins (greater than 80 to 90% sequence identity) can also be identified. The retrieved sequence are evaluated by mass accuracy of the peptides, matching of additional peptide masses in the MALDI spectrum after accounting for common modifications such as oxidation, acrylamidation of cysteine and missed cleavages and the use of secondary information (apparent isoelectric point and molecular weight). If any ambiguity about the identification by MALDI-MS still exists, the results must verified by an other identification method. Peptide mass fingerprint is generally carried out with a MALDI-TOF (Matrix-assisted laser desorption ionization time-of-flight ) instrument but can also be achieved ESI-TOF (Electrospray Ionisation time-of-flight) or LC-MS (Liquid Chromatography-Mass Spectrometry) mass spectrometer.
<tr><td>definition_reference:<td> PMID:10967324
<tr><td>definition_reference:<td> PMID:11752590
<tr><td colspan=2><hr><tr><tr><td>term:<td> full dna sequence: full identification by sequencing
<tr><td>id:<td> <a name="MI:0056">MI:0056
<tr><td>definition:<td> sequencing occurs during the course of the experiment.\nThe DNA to be sequenced is used as template for the in vitro synthesis, by DNA polymerase, of a set of partial replicas, all beginning at the same place, but terminating at different points along the DNA chain. The key to this method is the use of dideoxyribonucleoside triphosphates in which the deoxyribose 3'-OH group present in normal nucleotides is missing; when such a modified nucleotide is incorporated into a DNA chain, it blocks the elongation of the chain. To determine the full sequence, the four different chain-terminating dideoxyribonucleosides are used in competition with an excess of deoxyribonucleosides in separate DNA synthesis reactions on the same DNA template. When the products of these four reactions are analysed by electrophoresis in four parallel lanes of a denaturing polyacrylamide gel, the DNA sequence can be derived. Every lane displays a family of DNA fragments of different lengths, reflecting the different sites at which a specific residue occurs in the original DNA.
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> immunostaining
<tr><td>id:<td> <a name="MI:0422">MI:0422
<tr><td>definition:<td> immunostaining co-localisation
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> mass spectrometry
<tr><td>id:<td> <a name="MI:0427">MI:0427
<tr><td>definition:<td> mass spectrometry
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> monoclonal antibody
<tr><td>id:<td> <a name="MI:0072">MI:0072
<tr><td>definition:<td> monoclonal antibodies are monospecific antibodies produced in the supernatant of a cell line obtained by fusing a lymphocyte B to a myeloma cell line or selected by phage display technology.
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> ms protein sequence: de novo protein sequencing by mass spectrometry
<tr><td>id:<td> <a name="MI:0032">MI:0032
<tr><td>definition:<td> The strategy to determine the complete amino acid sequence of a protein by mass spectrometry relies on the generation of a nested set of fragments differing by one amino acid. This permits to reveal the identity of the residue that has been removed at each degradation step by measuring the mass difference of fragments differing of one residue. Peptide fragments can be obtained by protease treatment combined with the fragmentation promoted by collision (or other methods) within a tandem mass spectrometer. This approach can be carried out with LC MS/MS (Liquid Chromatography Tandem Mass Spectrometry), nanoESI MS/MS (nanoElectrospray Ionisation tandem mass spectrometry), or FTMS (Fourier Transform mass spectrometry) instruments.
<tr><td>definition_reference:<td> PMID:10984529
<tr><td colspan=2><hr><tr><tr><td>term:<td> nucleotide sequence: nucleotide sequence identification
<tr><td>id:<td> <a name="MI:0078">MI:0078
<tr><td>definition:<td> Identification of a nucleotide sequence.
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> partial dna sequence: partial dna sequence identification by hybridization
<tr><td>id:<td> <a name="MI:0080">MI:0080
<tr><td>definition:<td> Genes are recognised by hybridization of a probe with a fragment of the gene sequence.
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> partial id prot seq: partial identification of protein sequence
<tr><td>id:<td> <a name="MI:0433">MI:0433
<tr><td>definition:<td> partial protein sequence identification
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> peptide synthesis
<tr><td>id:<td> <a name="MI:0083">MI:0083
<tr><td>definition:<td> When one of the partners participates in the interaction with a relatively short peptide fragment, it is often convenient to precisely identify the minimal region that supports the interaction by synthesising a series of overlapping peptides and by testing them in the binding assay. Synthetic peptides that are identical with peptides synthesised in vivo are useful experimental tools for such studies. Peptides are routinely synthesised in a test tube from monomeric amino acids by condensation reactions that form peptide bonds. Peptides are constructed sequentially by coupling the C-terminus of a monomeric amino acid to the N-terminus of the growing peptide. To prevent unwanted reactions involving the amino groups and carboxyl groups of the side chains during the coupling steps, a protecting (blocking) group is attached to the side chains. Without these protecting groups, branched peptides would be generated. In the last steps of synthesis, the side chain-protecting groups are removed and the peptide is cleaved from the resin on which synthesis occurs.
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> polyclonal antibody
<tr><td>id:<td> <a name="MI:0086">MI:0086
<tr><td>definition:<td> Polyclonal antibodies are a mixture of different antibodies that represent the immune response, normally in an experimental animal, to any given antigen.
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> predetermined: predetermined participant
<tr><td>id:<td> <a name="MI:0396">MI:0396
<tr><td>definition:<td> Molecule whose sequence identity is not checked and has been assumed from external or previous experimental evidence(s).
<tr><td>definition_reference:<td> PMID:7940758
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> primer specific pcr
<tr><td>id:<td> <a name="MI:0088">MI:0088
<tr><td>definition:<td> Sequences can be identified in a DNA mixture by launching a PCR (Polymerase Chain Reaction) controlled by sequence specific primers. Such reaction starts only when the hybridization of the primer with a complementary sequence occurs.
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> protein sequence: protein sequence identification
<tr><td>id:<td> <a name="MI:0093">MI:0093
<tr><td>definition:<td> Single amino acid identification along a protein sequence.
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> sequence tag: sequence tag identification
<tr><td>id:<td> <a name="MI:0102">MI:0102
<tr><td>definition:<td> This approach leads to protein identification by combining mass measurement and short amino acid sequence information obtained by tandem mass spectrometry. This information is then used to automatically find the best match in a sequence database. A mixture of peptides derived from a protease digestion is analysed by nanoelectrospray LC-MS/MS (Liquid Chromatography Tandem Mass Spectrometer or nanoESI MS/MS) mass spectrometry. Electrospray mass spectrometry cannot be applied to dilute samples and is affected by high salt. As a consequence peptides, normally extracted from acrylamide gels by in situ proteolysis, are desalted and concentrated on a microcolumn followed by elution into a capillary used for nanoelectrospray tandem mass spectrometry. A first mass spectrum ("Normal mass spectrum" or "Q1 mass spectrum") gives information about the masses of all the peptides. Peptides observed in the normal mass spectrum are isolated in turn and dissociated into fragments by collision with gas molecules within the mass spectrometer. Some of the fragments obtained from a peptide constitute a nested set, differing by one amino acid, and the mass difference between them allows assignment of a partial sequence. The masses of the fragments define the position of the partial sequence in the peptide. Together with the cleavage specificity of the protease used to cleave the protein, and mass information such sequence tag provides much higher search specificity to match the a database entry. The procedure is repeated with several peptides from the digest, resulting in multiple identifications of the same protein or identification of several proteins from the peptide mixture. Unknown proteins can easily be identified by using the high specificity of the peptide sequence tag for searches in most sequence databases including EST or genome databases.
<tr><td>definition_reference:<td> PMID:10967324
<tr><td>definition_reference:<td> PMID:11752590
<tr><td colspan=2><hr><tr><tr><td>term:<td> southern blot
<tr><td>id:<td> <a name="MI:0103">MI:0103
<tr><td>definition:<td> A standard procedure to identify DNA fragments containing specific gene sequences. In this procedure i) a genome is fragmented using a restriction enzyme ii) the generated fragments are separated by electrophoresis iii) the fragments are transferred to a membrane iv)the membrane is incubated with a radio labelled probe that hybridises any complementary subsequence.
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr><tr><td>term:<td> western blot
<tr><td>id:<td> <a name="MI:0113">MI:0113
<tr><td>definition:<td> Western blot is a procedure to identify and quantify proteins. A mixture of protein is first submitted to an electrophoresis in denaturant condition and then electro-transferred from the gel to a membrane. The membrane is then incubated with a primary antibody specific for a given protein or a specific residue modification in the sample under analysis. A secondary antibody, radiolabelled or fused to a chromogenic enzyme, targets the first antibody and allows the visualisation of the protein band on the membrane.
<tr><td>definition_reference:<td> PMID:14755292
<tr><td colspan=2><hr><tr></table></html>