<html><table>
<tr><td>term:<td> de novo protein sequ: de novo protein sequencing by mass spectrometry
<tr><td>id:<td> <a name="MI:0032">MI:0032
<tr><td>definition:<td> The strategy to determine the complete amino acid sequence of a protein by mass spectrometry relies on the generation of a nested set of fragments differing by one amino acid. This permits to reveal the identity of the residue that has been removed at each degradation step by measuring the mass difference of fragments differing of one residue. Peptide fragments can be obtained by protease treatment combined with the fragmentation promoted by collision (or other methods) within a tandem mass spectrometer. This approach can be carried out with LC MS/MS (Liquid Chromatography Tandem Mass Spectrometry), nanoESI MS/MS (nanoElectrospray Ionisation tandem mass spectrometry), or FTMS (Fourier Transform mass spectrometry) instruments.
<tr><td>definition_reference:<td> PMID:10984529
<tr><td colspan=2><hr><tr><tr><td>term:<td> edman degradation
<tr><td>id:<td> <a name="MI:0039">MI:0039
<tr><td>definition:<td> In this procedure the N-terminus amino acid is cleaved from a polypeptide and identified by high-pressure liquid chromatography. The cycle is repeated on the evershortening polypeptide until all the residues are identified. On average only 20-30 consecutive cycles can be performed and lead to amino acid identification. Longer polypeptides or full length proteins must be cleaved by specific protease before Edman degradation and their sequences built by fragment overlapping.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> full identification: full identification by sequencing
<tr><td>id:<td> <a name="MI:0056">MI:0056
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> monoclonal antibody
<tr><td>id:<td> <a name="MI:0072">MI:0072
<tr><td>definition:<td> monoclonal antibodies are monospecific antibodies produced in the supernatant of a cell line obtained by fusing a lymphocyte B to a myeloma cell line or selected by phage display technology.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> nucleotide sequence: nucleotide sequence identification
<tr><td>id:<td> <a name="MI:0078">MI:0078
<tr><td>definition:<td> Identification of a nucleotide sequence.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> partial dna sequence: partial dna sequence identification by hybridization
<tr><td>id:<td> <a name="MI:0080">MI:0080
<tr><td>definition:<td> Genes are recognised by hybridization of a probe with a fragment of the gene sequence.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> participant detectio: participant detection
<tr><td>id:<td> <a name="MI:0002">MI:0002
<tr><td>definition:<td> Method to determine the proteins involved in the interaction.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> peptide massfingerpr: peptide massfingerprinting
<tr><td>id:<td> <a name="MI:0082">MI:0082
<tr><td>definition:<td> This approach leads to protein identification by matching peptide masses, as measured by mass spectrometry, to the ones calculated from in silico fragmentation of a protein sequence database. A peptide mixture from a tryptic digest is analysed by MALDI-MS (Matrix-assisted laser desorption ionization mass spectrometry). The list of peptide masses obtained by MALDI MS is automatically compared to the calculated masses of the predicted peptide fragments for each entry in the database. High mass accuracy is very important in order to obtain a statistically significant and unambiguous match This method is best applied to completely sequenced genomes and well characterised proteomes. However, depending on the data quality, proteins that are highly homologous to already characterised proteins (greater than 80 to 90% sequence identity) can also be identified. The retrieved sequence are evaluated by mass accuracy of the peptides, matching of additional peptide masses in the MALDI spectrum after accounting for common modifications such as oxidation, acrylamidation of cysteine and missed cleavages and the use of secondary information (apparent isoelectric point and molecular weight). If any ambiguity about the identification by MALDI-MS still exists, the results must verified by an other identification method. Peptide mass fingerprint is generally carried out with a MALDI-TOF (Matrix-assisted laser desorption ionization time-of-flight ) instrument but can also be achieved ESI-TOF (Electrospray Ionisation time-of-flight) or LC-MS (Liquid Chromatography-Mass Spectrometry) mass spectrometer.
<tr><td>definition_reference:<td> PMID:10967324
<tr><td>definition_reference:<td> PMID:11752590
<tr><td colspan=2><hr><tr><tr><td>term:<td> peptide synthesis
<tr><td>id:<td> <a name="MI:0083">MI:0083
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> polyclonal antibody
<tr><td>id:<td> <a name="MI:0086">MI:0086
<tr><td>definition:<td> Polyclonal antibodies are a mixture of different antibodies that represent the immune response, normally in an experimental animal, to any given antigen.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> predetermined
<tr><td>id:<td> <a name="MI:0396">MI:0396
<tr><td>definition_reference:<td> PMID:7940758
<tr><td colspan=2><hr><tr><tr><td>term:<td> primer specific pcr
<tr><td>id:<td> <a name="MI:0088">MI:0088
<tr><td>definition:<td> Sequences can be identified in a DNA mixture by launching a PCR (Polymerase Chain Reaction) controlled by sequence specific primers. Such reaction starts only when the hybridization of the primer with a complementary sequence occurs.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> protein sequence ide: protein sequence identification
<tr><td>id:<td> <a name="MI:0093">MI:0093
<tr><td>definition:<td> Single amino acid identification along a protein sequence.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> radio-labelled prote: radio-labelled protein
<tr><td>id:<td> <a name="MI:0517">MI:0517
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> sequence tag identif: sequence tag identification
<tr><td>id:<td> <a name="MI:0102">MI:0102
<tr><td>definition:<td> This approach leads to protein identification by combining mass measurement and short amino acid sequence information obtained by tandem mass spectrometry. This information is then used to automatically find the best match in a sequence database. A mixture of peptides derived from a protease digestion is analysed by nanoelectrospray LC-MS/MS (Liquid Chromatography Tandem Mass Spectrometer or nanoESI MS/MS) mass spectrometry. Electrospray mass spectrometry cannot be applied to dilute samples and is affected by high salt. As a consequence peptides, normally extracted from acrylamide gels by in situ proteolysis, are desalted and concentrated on a microcolumn followed by elution into a capillary used for nanoelectrospray tandem mass spectrometry. A first mass spectrum ("Normal mass spectrum" or "Q1 mass spectrum") gives information about the masses of all the peptides. Peptides observed in the normal mass spectrum are isolated in turn and dissociated into fragments by collision with gas molecules within the mass spectrometer. Some of the fragments obtained from a peptide constitute a nested set, differing by one amino acid, and the mass difference between them allows assignment of a partial sequence. The masses of the fragments define the position of the partial sequence in the peptide. Together with the cleavage specificity of the protease used to cleave the protein, and mass information such sequence tag provides much higher search specificity to match the a database entry. The procedure is repeated with several peptides from the digest, resulting in multiple identifications of the same protein or identification of several proteins from the peptide mixture. Unknown proteins can easily be identified by using the high specificity of the peptide sequence tag for searches in most sequence databases including EST or genome databases.
<tr><td>definition_reference:<td> PMID:10967324
<tr><td>definition_reference:<td> PMID:11752590
<tr><td colspan=2><hr><tr><tr><td>term:<td> southern blot
<tr><td>id:<td> <a name="MI:0103">MI:0103
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr><tr><td>term:<td> western blot
<tr><td>id:<td> <a name="MI:0113">MI:0113
<tr><td>definition:<td> Western blot is a procedure to identify and quantify proteins. A mixture of protein is first submitted to an electrophoresis in denaturant condition and then electro-transferred from the gel to a membrane. The membrane is then incubated with a primary antibody specific for a given protein or a specific residue modification in the sample under analysis. A secondary antibody, radiolabelled or fused to a chromogenic enzyme, targets the first antibody and allows the visualisation of the protein band on the membrane.
<tr><td>definition_reference:<td> PMID:INTACT
<tr><td colspan=2><hr><tr></table></html>